Avoid unnecessary allocations in Path.GetRandomFileName

justinvp opened the pull request:
Avoid unnecessary `byte[]`, `char[]`, and `string` allocations in `Path.GetRandomFileName`. A 10,000,000 iteration microbenchmark repeatedly calling `Path.GetRandomFileName()` shows a 3.8x reduction in gen0 GCs with a 55% speed improvement.

cc: @stephentoub, @bartonjs, @JeremyKuhne

JeremyKuhne commented on src/System.Runtime.Extensions/src/System/IO/Path.cs:
```diff
@@ -157,15 +157,19 @@ public static string GetFileNameWithoutExtension(string path)
             // rndCharArray is expected to be 16 chars
-            char[] rndCharArray = ToBase32StringSuitableForDirName(key).ToCharArray();
+            const int RndCharArrayLength = 16;
+            char* rndCharArray = stackalloc char[RndCharArrayLength];
```
Did you consider just creating a string of the right length and fixing it for your char*?

mellinoe replied:
I think the current way is preferable; it waits until the data is computed to construct the string and doesn't pin it at any point.

JeremyKuhne replied:
@mellinoe Creating a string from a char pointer will create the string, pin it, and copy the data over, with a bunch of additional bounds checks and a try catch. (See `String.CtorCharPtrStartLength`) If you created with char of null & count (`CtorCharCount`) it will simply be `FastAllocateString` and a pin here.

justinvp replied:
I went ahead and changed this to mutate the string before returning it per @JeremyKuhne's suggestion. That, along with removing the loop in `GetBase32CharsSuitableForDirName`, is 15% faster than what I had originally.

jamesqo replied:
@JeremyKuhne Said this in my comment below, but unfortunately mutating a string in-place is undefined behavior in C#: http://stackoverflow.com/a/31930863/4077294 I had some changes a while ago that were rejected b/c of this. We should just stick to the `stackalloc` version here.

JeremyKuhne replied:
@jamesqo You should never mess with a string that you didn't create or have passed on only because it is expected to be immutable. We know what the scope of the pin is here- and it is exactly what happens in the string class as per my comment above (see `CtorCharPtrStartLength`).

In general you don't want to go there. With native interop and using pointers here already I don't see a problem with it, particularly as we're in the Path class and it's methods often get hit in tight loops. That said, this particular method shouldn't be called like that, and some others on the team might feel differently. :)

jamesqo replied:
@JeremyKuhne I see your point-- we allocate the string, so we know that it's "safe" to modify it so long as it isn't reachable outside of the method. Still, I'd at least hold off merging this PR until some others weigh in on this.

mellinoe replied:
I am surprised to hear that creating the string, pinning it, and mutating it is faster than the stackalloc ->ctor implementation (even if it pins internally). That seems like something that should be fixed to me. Stackalloc->ctor seems like the "proper" way to do this.

JeremyKuhne replied:
@mellinoe It is because of the extra validation, copy, and try catch I assume. :)

mellinoe replied:
Ahh, that darn validation is always slowing things down :smile:

JeremyKuhne commented on src/System.Runtime.Extensions/src/System/IO/Path.cs:
```diff
@@ -157,15 +157,19 @@ public static string GetFileNameWithoutExtension(string path)
         // Returns a cryptographically strong random 8.3 string that can be 
         // used as either a folder name or a file name.
-        public static string GetRandomFileName()
+        public static unsafe string GetRandomFileName()
```
I'd move the unsafe inside the method here as this is public.

stephentoub replied:
unsafe doesn't affect the signature

JeremyKuhne commented:
Thanks! I made a couple comments, but generally looks good.

bartonjs commented on src/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.RAND.cs:
```diff
@@ -8,8 +8,16 @@ internal static partial class Interop
     {
+        internal static unsafe bool GetRandomBytes(byte[] buf, int num)
+        {
+            fixed (byte* pBuf = buf)
```
Debug.Assert(num <= buf.Length)

bartonjs replied:
Oh, and that it's positive.  That'd be good, too :)

justinvp replied:
Fixed.

stephentoub commented on src/System.Runtime.Extensions/src/System/IO/Path.cs:
```diff
@@ -356,41 +360,42 @@ private static string CombineNoChecks(string path1, string path2, string path3)
+            Debug.Assert(chars != null);
+            // This routine is optimized to be used with bytes of length 10 and chars of length 16.
+            Debug.Assert((byteCount >= 5) && ((byteCount % 5) == 0), $"Unexpected {nameof(byteCount)}");
+            Debug.Assert((charCount >= 8) && ((charCount % 8) == 0), $"Unexpected {nameof(charCount)}");
```
Since this is only called from the one location, should we just assert what's called out in the comment rather than the more complicated asserts currently there?

justinvp replied:
Done. I'm now asserting the exact lengths needed for the 12 length string needed by `GetRandomFileName` (byteCount == 9, charCount == 12), and since I did that, I went ahead and removed the loop completely.

JeremyKuhne commented:
Cool- looks good to me, thanks for the fine tuning!

jamesqo commented on src/System.Runtime.Extensions/src/System/IO/Path.WinRT.cs:
```diff
@@ -11,12 +12,18 @@ namespace System.IO
+
+            Debug.Assert(arr.Length == byteCount);
+
+            Marshal.Copy(arr, 0, new IntPtr(bytes), byteCount);
```
Won't this lead to a perf regression for WinRT? Before it was only making one copy (from the `IBuffer` to the byte array), now it's making two.

justinvp replied:
If you know of a more efficient way, let me know. I don't think it's a net regression due to the other `char[]` and `string` allocation reductions elsewhere and the small array size.

One idea I had was to query to see if the result of `CryptographicBuffer.GenerateRandom` implements `IBufferByteAccess` and copy the data directly instead of allocating a temp array via `CryptographicBuffer.CopyToByteArray`. However, I'm not familiar with building/testing the WinRT build of this library and I'm not sure if the result actually is an `IBufferByteAccess`. I suppose I could build a simple UWP app to try this out.

jamesqo replied:
@justinvp I don't think it's a good idea to add more complexity to this method (the byte array in question is only length 9, so the type cast itself could be more expensive). The only way I see to avoid these allocations is to have a different version of `GetRandomFileName` for each platform; it's unfortunate you can't do `stackalloc` for the parent function, otherwise you could write something like

```cs
// Path.cs
public static string GetRandomFileName()
{
    byte* pointer; // null on platforms where we allocate an array
    byte[] array; // null on platforms where we use stackalloc
    GetRandomCryptoBytes(9, out pointer, out array);
    if (pointer != null) DoWorkWith(pointer);
    else fixed (byte* pArray = array) DoWorkWith(pointer);
}

// Path.Windows.cs
static void GetCryptoRandomBytes(int byteCount, out byte* pointer, out byte[] unused)
{
    unused = null;
    pointer = stackalloc byte[byteCount]; // won't work, unfortunately
}

// Path.WinRT.cs
static void GetCryptoRandomBytes(int byteCount, out byte* unused, out byte[] array)
{
    unused = null;
    array = new byte[byteCount];
}
```

For now you should probably keep it the way it is.

justinvp replied:
I tried the following in a UWP app...

```csharp
private static unsafe void GetCryptoRandomBytes(byte* bytes, int byteCount)
{
    Debug.Assert(bytes != null);
    Debug.Assert(byteCount > 0);

    IBufferByteAccess buffer = (IBufferByteAccess)CryptographicBuffer.GenerateRandom((uint)byteCount);
    byte* pBuffer = (byte*)buffer.GetBuffer();
    for (int i = byteCount - 1; i >= 0; i--)
    {
        bytes[i] = pBuffer[i];
    }
}

[ComImport]
[Guid("905a0fef-bc53-11df-8c49-001e4fc686da")]
[InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
internal interface IBufferByteAccess
{
    IntPtr GetBuffer();
}
```

A 1,000,000 microbenchmark shows that this is 4.8x faster than what I had originally.

jamesqo replied:
@justinvp Ah, ok. I hadn't realized that the result of `GenerateRandom` would always be castable to `IBufferByteAccess` (and that `GetBuffer` would return IntPtr instead of byte[]). In that case, I'd say it's ok to make the change.

justinvp replied:
@jamesqo, FYI, I re-ran the microbenchmark a few times in a UWP app and the speed is actually the same (only a hair faster), just without the extra byte[] allocation.

jamesqo commented on src/System.Runtime.Extensions/src/System/IO/Path.cs:
```diff
@@ -157,17 +157,21 @@ public static string GetFileNameWithoutExtension(string path)
+
+            const int RandomFileNameLength = 12;
+            string randomFileName = new string('\0', RandomFileNameLength);
+            fixed (char* pRandomFileName = randomFileName)
```
It's supposed to be UB to modify the characters of a string; you should instead `stackalloc` an array (as `RandomFileLength` is relatively small) and copy the data over to the string.

JeremyKuhne replied:
That is where he started. It is fine to do here as we haven't passed the string to anyone else yet. It is significantly more efficient to locally do what the constructor for string from char* is doing anyway.

justinvp replied:
What does "UB" stand for? I originally implemented it the way you suggest, but @JeremyKuhne suggested mutating the string. I don't have a strong preference either way.

jamesqo replied:
@justinvp UB -> undefined behavior; I had some changes a while ago that were rejected since they involved modifying a string in-place ([here](https://github.com/dotnet/corefx/pull/2694#discussion_r36587700)). See also this StackOverflow answer: http://stackoverflow.com/a/31930863/4077294

ellismg replied:
FWIW, I would also propose the stackalloc thing.  I understand that there's an addition copy when you create the string, but unless performance traces showed that copy was killing us, I'd rather write the "safer" version of the code, just to encourage good habits.

justinvp commented:
Thanks for the feedback, everyone. I pushed three new commits:

 1. Reverted back to using the stackallocated `char*` and `string` constructor instead of mutating a fixed `string`.

 2. Fixed the Path.WinRT.cs build break due to missing `System.Diagnostics` namespace import.

 3. Avoid an unnecessary `byte[]` allocation in `Path.GetRandomFileName` on WinRT. The implementation of WinRT's private `Path.GetCryptoRandomBytes` uses `CryptographicBuffer.GenerateRandom` to generate the random bytes, which returns an `IBuffer`. It then calls `CryptographicBuffer.CopyToByteArray` to create a new `byte[]` from the `IBuffer`. We can avoid the `byte[]` allocation by casting the `IBuffer` to [`IBufferByteAccess`](https://msdn.microsoft.com/en-us/library/windows/desktop/hh438362(v=vs.85).aspx), which allows direct access to the bytes.

  I tested the new WinRT implementation in a one-off UWP app as I haven't been able to run the tests locally with `/p:Configuration=Windows_netcore50aot_Debug`. Attempting to do so fails with the following error:

  ```
C:\dev\corefx\src\System.Runtime.Extensions\tests>msbuild /t:BuildAndTest /p:Configuration=Windows_netcore50aot_Debug

... lots of msbuild output removed...

Running tests... Start time:  0:32:08.66
Command(s):
CoreRun.exe xunit.console.netcore.exe System.Runtime.Extensions.Tests.dll  -xml testResults.xml -notrait Benchmark=true  -notrait category=OuterLoop -notrait category=failing -notrait category=nonwindowstests

Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Private.CoreLib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
   at Xunit.ConsoleClient.Program.Main(String[] args)
```

  Could someone help validate the WinRT changes?

stephentoub commented:
> We can avoid the  byte[]  allocation by casting the  IBuffer  to IBufferByteAccess , which allows direct access to the bytes.

@justinvp, I suggest limiting the changes in this PR for the WinRT implementation to the bare minimum to enable the rest of the change to work and go through.  This can then be optimized further when there's better infrastructure available for validating the change.

stephentoub commented on src/System.Runtime.Extensions/src/System/IO/Path.cs:
```diff
@@ -356,58 +357,58 @@ private static string CombineNoChecks(string path1, string path2, string path3)
+            chars[1] = s_Base32Char[b1 & 0x1F];
+            chars[2] = s_Base32Char[b2 & 0x1F];
+            chars[3] = s_Base32Char[b3 & 0x1F];
+            chars[4] = s_Base32Char[b4 & 0x1F];
```
Nit: As long as you're already modifying all of the consumption sites, consider renaming ```s_Base32Char``` to ```s_base32Char```.

justinvp replied:
Done.

stephentoub commented on src/System.Runtime.Extensions/src/System/IO/Path.cs:
```diff
@@ -356,58 +357,58 @@ private static string CombineNoChecks(string path1, string path2, string path3)
                 'y', 'z', '0', '1', '2', '3', '4', '5'};
 
-        private static string ToBase32StringSuitableForDirName(byte[] buff)
+        private static unsafe void GetBase32CharsSuitableForDirName(byte* bytes, int byteCount, char* chars, int charCount)
```
Nit: Given the change in implementation, I'm not sure this name is still the best one.  How about something like Populate83FileNameFromRandomBytes, or something like that.  The ```pRandomFileName[8] = '.'``` could then be moved into the function, instead of the code currently setting the 8th char.

justinvp replied:
Done.

stephentoub commented on src/System.Runtime.Extensions/src/System/IO/Path.cs:
```diff
@@ -356,58 +357,58 @@ private static string CombineNoChecks(string path1, string path2, string path3)
 
-            // For every 5 bytes, 8 characters are appended.
-            StringBuilder sb = StringBuilderCache.Acquire();
+            // This routine is optimized to be used with bytes of length 9 and chars of length 12.
```
Not just optimized: it requires it now.

justinvp replied:
Yup, fixed the comment.

stephentoub commented on src/System.Runtime.Extensions/src/System/IO/Path.cs:
```diff
@@ -356,58 +357,58 @@ private static string CombineNoChecks(string path1, string path2, string path3)
 
-            // Create l char for each of the last 5 bits of each byte.  
+            // For every 5 bytes, 8 characters are appended.
+            // Create l char for each of the last 5 bits of each byte.
```
This comment is now also a bit misleading.

justinvp replied:
Fixed.

justinvp commented:
> I suggest limiting the changes in this PR for the WinRT implementation to the bare minimum to enable the rest of the change to work and go through.

@stephentoub, sounds good to me. I reverted the WinRT changes back to how I originally had it and have addressed your other feedback.

justinvp commented:
Test Innerloop CentOS7.1 Debug Build and Test please (#8606)
Test Innerloop Ubuntu14.04 Debug Build and Test please (#8606)

justinvp commented:
Test Innerloop CentOS7.1 Debug Build and Test please (#8633)

stephentoub commented:
Looks good.  Thanks for the improvement, @justinvp.
